[bug] handle case when remote peer closes connection

c-p-i-o opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145759
* __->__ #145757
* #145756

Summary:
In the case where remote peer closes the connection, nread returns 0. In
this case, we still want to free up the allocated buffer.
Also, reorder the if so that the likely success cases (nread > 0) is at
the top of the function with an early return.

Test Plan:
unit tests

Differential Revision: [D68733192](https://our.internmc.facebook.com/intern/diff/D68733192)

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

c-p-i-o committed:
[bug] handle case when remote peer closes connection

Summary:
In the case where remote peer closes the connection, nread returns 0. In
this case, we still want to free up the allocated buffer.
Also, reorder the if so that the likely success cases (nread > 0) is at
the top of the function with an early return.

Test Plan:
unit tests

[ghstack-poisoned]

XilunWu reviewed (commented):
Will using `if-elif` and getting rid of `return` be better?

c-p-i-o commented:
> Will using `if-elif` and getting rid of `return` be better?

It's a style thing. [early return principle](https://dev.to/shameel/-clean-coding-tip-early-return-principle-47kj)

This style often competes with [single return](https://www.anthonysteele.co.uk/TheSingleReturnLaw)


If you prefer the if-elif i.e. `single return`, I can change it. Let me know!

c-p-i-o committed:
Update on "[bug] handle case when remote peer closes connection"


Summary:
In the case where remote peer closes the connection, nread returns 0. In
this case, we still want to free up the allocated buffer.
Also, reorder the if so that the likely success cases (nread > 0) is at
the top of the function with an early return.

Test Plan:
unit tests

[ghstack-poisoned]

XilunWu commented on torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp:
```diff
@@ -116,24 +116,27 @@ class UvTcpSocket : public UvHandle {
         C10D_WARNING("Error processing client message: {}", ex.what());
         uv_socket->close();
       }
+    } else if (nread <= 0) {
```
nit: we can just use `else` here

XilunWu reviewed (approved):
LGTM! Let's also import to internal to test before landing.

c-p-i-o committed:
Update on "[bug] handle case when remote peer closes connection"


Summary:
In the case where remote peer closes the connection, nread returns 0. In
this case, we still want to free up the allocated buffer.
Also, reorder the if so that the likely success cases (nread > 0) is at
the top of the function with an early return.

Test Plan:
unit tests

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68733192).

c-p-i-o commented:
> LGTM! Let's also import to internal to test before landing.

Imported! Will wait for internal tests to pass.

c-p-i-o commented:
@pytorchbot merge
